src/net/tcp.ts

```diff
@@ -116,6 +111,18 @@ export class Server {
       // TODO: investigate which other values occur
       this.fulfillShutdown_(SocketCloseKind.UNKOWN);
     }
+
+    // Since there may be other calls to this provider in progress which will
+    // fail to resolve if we destroy the communication channel now, postpone
+    // doing so until the next iteration of the event loop.
+    setTimeout(() => {
```

Comment:
OK, it was kinda fun to try to come up with a non-horrible solution...I'm not sure I have, although the integration tests pass now.

This of course is without changing the guarantee on when `onDisconnect` fires, which we might also consider.